evm/src/cpu/kernel/asm/main.asm

```diff
@@ -14,14 +14,20 @@ global main:
     // Initialise the shift table
     %shift_table_init
 
-    // Second, load all MPT data from the prover.
-    PUSH hash_initial_tries
-    %jump(load_all_mpts)
-
 global hash_initial_tries:
-    %mpt_hash_state_trie   %mload_global_metadata(@GLOBAL_METADATA_STATE_TRIE_DIGEST_BEFORE)    %assert_eq
+    // We compute the length of the trie data segment in `mpt_hash` so that we
+    // can check the value provided by the prover.
+    // We initialize the segment length with 1 because the segment contains 
+    // the null pointer `0` when the tries are empty.
+    PUSH 1
+    %mpt_hash_state_trie  %mload_global_metadata(@GLOBAL_METADATA_STATE_TRIE_DIGEST_BEFORE)    %assert_eq
+    // stack: trie_data_len
     %mpt_hash_txn_trie     %mload_global_metadata(@GLOBAL_METADATA_TXN_TRIE_DIGEST_BEFORE)      %assert_eq
+    // stack: trie_data_len
     %mpt_hash_receipt_trie %mload_global_metadata(@GLOBAL_METADATA_RECEIPT_TRIE_DIGEST_BEFORE)  %assert_eq
+    // stack: trie_data_full_len
+    %mload_global_metadata(@GLOBAL_METADATA_TRIE_DATA_SIZE)
+    %assert_eq
```

Comment:
> I can remove the length and pointers for the `extra_block_data` and replace that with a memory store in the Kernel, if that' s what you mean? (I still need to have a memory read since `GlobalMetadata` cannot yet be preinitialized given the constraints).

Yeah removing them from `extra_block_data` would remove complexity. But also, I don't really understand why we need to check the trie data length here.

Reply:
Can't there be an issue if, say, we set nondeterministically the length to 0 (but still have a nonempty trie)? `mpt_hash` doesn't use the trie data size, so it could lead to a nonempty hash, with the rest of the code assuming an empty trie data segment, right?

Reply:
Ah right, makes sense thanks!